tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -274,6 +368,65 @@ std::vector<TensorView*> mha(
   return {qkv, sdpa_output, linear2, dropout_result};
 }
 
+// TODO: These linear_backwards helper functions can be merged once
+// we do not have logically split rfactor domain.
+struct LinearBackwardsResult {
+  TensorView* grad_x;
+  TensorView* grad_w;
+  TensorView* grad_b;
+};
+
+// x format: [i0, i1] dtype
+// weight format: [DID(D), i1, i2/D] dtype
+// grad format: [DID(D) i0, i2/D] float
+// outputs: grad_x [i0, i1] dtype
+// grad_w [DID i1, i2/D] dtype
+// grad_b [DID i2/2] float
+LinearBackwardsResult linear_backwards(
+    TensorView* x,
+    TensorView* w,
+    TensorView* grad,
+    DataType dtype) {
```

Comment:
Can we use `x->dtype()` instead? (One less parameter to worry about)

Reply:
yes. will update all the helpers!